### ISSUE TITLE
preferEmbeddedResources ditching templatePraameters?

### ISSUE THREAD

**Issue by enrique-ramirez** [traverson/traverson-hal#29]
What's the best way to:
a) Prefer embedded resources as to minimize HTTP requests...
b) ...while still being able to ask for specific items using link params?

Extended explanation: For an API whose replies look like this:

```json
// ./orders paginated by 10 per page
{
  "count": 11,
  "_embedded": {
    "my:order": [
      {"id": "1", "name": "Order 1"},
      {"id": "2", "name": "Order 2"},
      {"id": "3", "name": "Order 3"}
    ]
  },
  "_links": {
    "self": "/orders",
    "my:order": {
      "url": "/orders{/id}",
      "templated": true
    }
  }
}
```

```json
// ./orders{/id}
{
  "id": "1",
  "name": "Some name",
  "_links": {
    "self": "/orders/1"
  }
}
```

If I do this:

```javascript
traverson
  .from('/')
  .jsonHal()
  .newRequest()
  .follow(['my:orders', 'my:order'])
  .withTemplateParameters({ id: 11 })
  .get()
```

Everything works fine. Traverson does a request for root, then for `my:orders` rel, then finally to `my:order` applying the `id` parameter to the templated URL.

However, if I add the `.preferEmbeddedResources()` option like this:

```javascript
traverson
  .from('/')
  .jsonHal()
  .newRequest()
  .preferEmbeddedResources()
  .follow('my:orders', 'my:order')
  .withTemplateParameters({ id: 11 })
  .get()
```

It will ignore the template parameters. It will do a request for root, then for `my:orders` and then return the first element that matches the `my:order` key in the `_embedded` response for the collection.

I kinda get why this is. By telling it that I prefer embedded resources, it'll try to filter whatever the last rel returns (in this case, an array). If I added something like `my:order[3]` it'll return the third element of the array since that's what I'm asking it to filter. It still took my a bit by surprise and I believe that, if this is expected behaviour, maybe it could be better explained somewhere? (or maybe I'm just stupid and couldn't find the proper doc).

**Comment by basti1302** [comment]
Hey, sorry for the late reply. I didn't yet reproduce your exact case, but I think you would need to do something like

```
...
.preferEmbeddedResources()
.follow(['my:orders', 'my:order[id:11]')
.withTemplateParameters({ id: 11 })
....
```

Note the secondary key I used in the second link rel: my:order**[id:11]**
Could you try that and report the result back? I think it should work, but I haven't tested it.

Some explanations:

* `withTemplateParameters` is a Traverson core method, it is not traverson-hal specific. It is not integrated at all with HAL-semantics (and is not meant to be). It just says: Whenever you encounter an URI-template, use the following parameter substitutions.
* `.follow('my:orders', 'my:order')` is actually underspecified. It just tells the HAL-machinery
    * Go to the orders resource (there is only one, so this is fine)
    * Next, go to the order resource. But hey, there are multiple orders, so which one do you want?
* I there are multiple resources and you do not specify which one you want, traverson-hal chooses the first one arbitrarily.
* So, in the preferEmbeddedResources case, HAL finds embedded orders, and since you didn't say which order you want, the first from the array is considered to be a the result you wanted, so it is returned. 
* In the  non-preferEmbeddedResources case, both the linked and the embedded are found but the linked resource is preferred. The fact that you get the one with id 11 in this case is a bit of random luck, because the UIR template parameter is indeed applied before making the request. 

So, in conclusion, it would be much cleaner to tell the HAL machinery that you are looking for my:order[id:11] and then everything should work.

Hope that helps!

**Comment by basti1302** [comment]
Actually, I just realized, that this might not work for your use case. With my:order[id:11], when there is no such embedded element present, it would probably also not use the link from `_links` to fetch  /orders/11. For that, the links would need to look like this:

```
  "_links": {
    "my:order": [{
      "url": "/orders/1",
      "id": 1
    }, {
      "url": "/orders/2",
      "id": 2
    }, 
    ...
    }, {
      "url": "/orders/11",
      "id": 11
    }]
  }
```

Only then would traverson-hal be able to select the correct link from the array of links with the "my:order" link relation. I'm pretty sure this is not what you want, even if you have control over the backend.

So currently I have no idea how to solve your use case. The complications come from the fact that URI templates and HAL are not integrated and do not interact with each other in any meaningful way. I'm not sure if they should, and if so, how.

If you have any suggestions, please let me know.

**Comment by basti1302** [comment]
Also, this secondary key business had been mentioned in the docs, but maybe not as detailed as it should be. I updated the docs, see https://github.com/traverson/traverson-hal#selecting-from-a-collection-of-resources / [diff](https://github.com/traverson/traverson-hal/commit/bdeb0d24e26aaa1360ebcd4c46288d480d4defa2).

**Comment by enrique-ramirez** [comment]
Hi @basti1302 !

So secondary keys are for filtering the server response while template parameters are used to ask the server for specific stuff.

Ok, so let's talk about an ideal world, shall we?

Let's say I do this:

```
traverson
  .from('/')
  .jsonHal()
  .newRequest()
  .preferEmbeddedResources()
  .follow('my:orders', 'my:order[id:11]')
  .withTemplateParameters({ id: 11 })
  .get()
```

Which is basically the same thing I was doing, but with the addition of the `[id:11]` secondary key. I'm not entirely happy with the syntax, since there's duplicated information (id: 11) in 2 different places, but I understand that they represent completely different things and the fact that they are linked in this scenario is mere coincidence. So let's go ahead with that.

What I think should happen is this:

1. Traverson GETs root.
2. Traverson GETs `my:orders` rel.
3. Traverson finds an `_embedded` key in the response, it begins searching for the object with the key `id` with value `11`, as per the secondary key instructions.

Here we have two scenarios: it finds the item in `_embedded` or it does not. The former is currently implemented and working as expected. It's the latter that seems incomplete to me. So continuing from the failure scenario:

4. Traverson then goes to `_links` and finds a `my:order` rel, which is a templated URL which happens to accept the parameters I specified in `withTemplateParameters`.
5. Traverson applies the parameters to the templated URL and after receiving the response, it looks again for an item with a key `id` with value `11`.
6. If traverson finds it, then it returns it and end of story. If it does not, it throws an error.

I think that is how it should work. 

Now, I believe the problem comes from the fact that secondary keys filtering is also applied to links, so Traverson would look for a link with a key `id` with value `11`, not find it and then error.

I believe the main issue is with the `withTemplateParameters` function, so maybe this is a change for `traverson` instead of traverson hal?

Say I want to `follow(['my:albums', 'my:album'])`. Say also that both rels are templated and both of them accept an `id` parameter. How would I go about with this? I think the answer would be to do a single `follow(['my:albums']).withTemplateParameters({id: artistId})` and then use the `traversal` object from the response to continue to `follow(['my:album']).withTemplateParameters({id: albumId})`.

This is a dumb example, but you get the idea: if I want to apply different template parameters to the different rels on the same traversal, I can't.

I'm not entirely sure if `withTemplateParameters` attempts to apply the parameters to ALL links in the traversal or just the last one, but either case seems wrong. The way template parameters are being applied currently seems to be a bit too late.

Maybe what we need is a new syntax? Something like:

```javascript
follow([
    {rel: 'my:albums', params: {id: 1}},
    {rel: 'my:album', params: {id: 2}},
])
```
?

This way, traverson could do the GETs with the specified parameters FIRST, then attempt secondary key filtering AFTER, when the response is what the developer intended.

With this addition, I believe we could solve all problems:

* It wouldn't be a breaking change. Strings on the follow array would continue to work as expected.
* Secondary key filtering would still work the same.
* Applying different template parameters to different rels would be easy as pie.
* More to the point of this ticket: *Traverson could do a GET first, and secondary key filtering second*. This would solve the problem for embedded resources, I believe.

What do you think?

**Comment by enrique-ramirez** [comment]
Or maybe I am overthinking this and we can do this without a new syntax. Basically, have the process exactly the same as it currently is, but when it gets to the point of "I can't find the element `my:order[id:11]` in neither `_embedded` nor `_links`" have it think: "but I have template params. Lemme attempt one last GET and see if `my:order[id:11]` is there"

**Comment by basti1302** [comment]
Sorry this totally fell through the cracks.

I just revisited it again. However, I think the current behaviour works as designed and there is no good motivation to change it.

> What I think should happen is this:
> 
> Traverson GETs root.
> Traverson GETs my:orders rel.
> Traverson finds an _embedded key in the response, it begins searching for the object with the key id with value 11, as per the secondary key instructions.
> Here we have two scenarios: it finds the item in _embedded or it does not. The former is currently implemented and working as expected. It's the latter that seems incomplete to me. So continuing from the failure scenario:
>
> Traverson then goes to _links and finds a my:order rel, which is a templated URL which happens to accept the parameters I specified in withTemplateParameters.

No, it would not find a link with rel `my:order` because you asked for a link with rel `my:order[id:11]` and there is no such link. It does not matter that there _would_ be an existing resource at URL `/orders/11`. It will only follow a link when the `rel` matches.  

> Traverson applies the parameters to the templated URL and after receiving the response, it looks again for an item with a key id with value 11.
> If traverson finds it, then it returns it and end of story. If it does not, it throws an error.
> I think that is how it should work.

This would be an interesting mechanism, however, it is quite different then what the HAL specification says. If you want that behaviour, you could write your own media type plugin for Traverson (possibly a for of traverson-hal with the required modifications - that might sound more challenging than it actually is). Check out <https://github.com/traverson/traverson/blob/master/user-guide.markdown#using-media-type-plug-ins> and <https://github.com/traverson/traverson/blob/master/user-guide.markdown#implementing-media-type-plug-ins>.

>I believe the main issue is with the withTemplateParameters function, so maybe this is a change for traverson instead of traverson hal?
>
>Say I want to follow(['my:albums', 'my:album']). Say also that both rels are templated and both of them accept an id parameter. How would I go about with this? I think the answer would be to do a single follow(['my:albums']).withTemplateParameters({id: artistId}) and then use the traversal object from the response to continue to follow(['my:album']).withTemplateParameters({id: albumId}).

This is easily possible, see https://github.com/traverson/traverson/blob/master/user-guide.markdown#uri-templates, in particular the part beginning with "Instead of using a single object to provide the template parameters for each step, you can also provide an array of objects."

**Comment by basti1302** [comment]
Since a lot of time has passed since the initial discussion (totally my fault) I assume that interest in this issue might have lessened, so to keep the issue list clean, I'll go ahead and close this. Feel free to comment or reopen, though.

**Comment by enrique-ramirez** [comment]
@basti1302 thank you for your reply! I have moved on from this issue, indeed, but I really appreciate the thorough reply. I now understand more of the process and have a clearer idea of what I should be doing instead.
